File: server/functions/src/accessors/external_apps/external_app_telegram_bot_accessor.ts
import {SemurEngineConfig} from "../../config";
import {SemurEngineErrorBuilder} from "../../semur_engine/semur_engine";
import {TelegramBotIntegration} from "../../models.pb/external_apps/telegram_bot_integration";

export enum ExternalAppTelegramBotQueryKey {
    All = "all",
}


// User query arguments from query key
export interface ExternalAppTelegramBotQueryArgs {
    [ExternalAppTelegramBotQueryKey.All]: Record<string, never>;
}

export class ExternalAppTelegramBotAccessor {
    static firestoreCollectionName = "external_app_telegram_bot";
    static firestoreCollection = SemurEngineConfig.db.collection(this.firestoreCollectionName);

    static async exists(userId: string): Promise<boolean> {
        const snapshot = await this.firestoreCollection.doc(userId).get();
        return snapshot.exists;
    }

    static async get(integrationId: string): Promise<TelegramBotIntegration> {
        const snapshot = await this.firestoreCollection.doc(integrationId).get();
        if (!snapshot.exists) {
            throw SemurEngineErrorBuilder.notFound("integration not found");
        }
        return TelegramBotIntegration.fromJSON(snapshot.data() as Record<string, unknown>);
    }

    static async set(integration: TelegramBotIntegration): Promise<void> {
        await this.firestoreCollection.doc(integration.id).set({...integration});
    }

    static async delete(integrationId: string): Promise<void> {
        await this.firestoreCollection.doc(integrationId).delete();
    }

    static async update(integration: TelegramBotIntegration): Promise<void> {
        await this.firestoreCollection.doc(integration.id).update({...integration});
    }

    static async updateCustomFields(userId: string, data: { [key: string]: unknown }): Promise<void> {
        await this.firestoreCollection.doc(userId).update(data);
    }

    static async query(
        queryKey: ExternalAppTelegramBotQueryKey,
        args: ExternalAppTelegramBotQueryArgs[ExternalAppTelegramBotQueryKey] = {},
    ): Promise<TelegramBotIntegration[]> {
        const query = await this.buildQuery(queryKey, args);
        const querySnapshot = await query.get();
        return querySnapshot.docs.map((doc) => TelegramBotIntegration.fromJSON(doc.data() as Record<string, unknown>));
    }

    // Build query given query key and arguments
    static async buildQuery(
        queryKey: ExternalAppTelegramBotQueryKey,
        args: ExternalAppTelegramBotQueryArgs[ExternalAppTelegramBotQueryKey] = {},
    ): Promise<FirebaseFirestore.Query> {
        if (!this.firestoreCollection) {
            throw SemurEngineErrorBuilder.internalError("Firebase collection is not defined");
        }
        switch (queryKey) {
            case ExternalAppTelegramBotQueryKey.All:
                return this.firestoreCollection;
            default:
                throw SemurEngineErrorBuilder.internalError(`Unknown query key: ${queryKey}`);
        }
    }
}

File: client/semur/lib/accessors/external_apps/external_app_telegram_bot_accessor.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:semur/config/application.dart';
import 'package:semur/models.pb/external_apps/telegram_bot_integration.pbserver.dart';
import 'package:semur/transformers/external_apps/telegram_bot_integration_firebase_transformer.dart';

enum ExternalAppTelegramBotAccessorQueryKey { all }

Map<ExternalAppTelegramBotAccessorQueryKey, List<String>>
externalAppTelegramBotAccessorQueryArguments = {
  ExternalAppTelegramBotAccessorQueryKey.all: [],
};

class ExternalAppTelegramBotAccessor {
  static String firebaseCollectionName = "external_app_telegram_bot";

  final FirebaseFirestore firestore;
  ExternalAppTelegramBotAccessor({required this.firestore});

  Query buildQuery({
    required ExternalAppTelegramBotAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
  }) {
    // Validate required arguments
    List<String> requiredArgs =
        externalAppTelegramBotAccessorQueryArguments[queryKey] ?? [];
    for (String arg in requiredArgs) {
      if (!arguments.containsKey(arg)) {
        throw Exception("Missing required argument '$arg' for query $queryKey");
      }
    }

    // Build query based on key
    switch (queryKey) {
      case ExternalAppTelegramBotAccessorQueryKey.all:
        return firestore.collection(firebaseCollectionName);
    }
  }

  Future<List<TelegramBotIntegration>> getQuery({
    required PermissionRoles callerRole,
    required ExternalAppTelegramBotAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
    int? limit,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];

    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }

    try {
      Query query = buildQuery(queryKey: queryKey, arguments: arguments);
      if (limit != null) {
        query = query.limit(limit);
      }
      QuerySnapshot snapshot = await query.get();
      return snapshot.docs
          .map(
            (doc) =>
                TelegramBotIntegrationFirebaseTransformer.fromFirebase(doc),
          )
          .toList();
    } catch (e) {
      throw Exception(
        "Error fetching telegram bot integrations: ${e.toString()}",
      );
    }
  }

  Future<int> countQuery({
    required PermissionRoles callerRole,
    required ExternalAppTelegramBotAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];

    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }

    try {
      AggregateQuerySnapshot snapshot =
          await buildQuery(
            queryKey: queryKey,
            arguments: arguments,
          ).count().get();
      return snapshot.count ?? 0;
    } catch (e) {
      throw Exception(
        "Error fetching telegram bot integrations: ${e.toString()}",
      );
    }
  }

  Future<TelegramBotIntegration> get({
    required String integrationId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }
    try {
      DocumentSnapshot<Map<String, dynamic>> snapshot =
          await firestore
              .collection(firebaseCollectionName)
              .doc(integrationId)
              .get();

      if (!snapshot.exists) {
        throw Exception("External App not found.");
      }
      return TelegramBotIntegrationFirebaseTransformer.fromFirebase(snapshot);
    } catch (e) {
      throw Exception(
        "Error fetching telegram bot integration: ${e.toString()}",
      );
    }
  }

  Future<void> set({
    required TelegramBotIntegration integration,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to save this data.");
    }
    try {
      await firestore
          .collection(firebaseCollectionName)
          .doc(integration.id)
          .set(
            TelegramBotIntegrationFirebaseTransformer.toJson(integration),
            SetOptions(merge: true),
          );
    } catch (e) {
      throw Exception("Error saving telegram bot integration: ${e.toString()}");
    }
  }

  Future<void> delete({
    required String integrationId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [PermissionRoles.admin];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to delete this data.");
    }
    try {
      await firestore
          .collection(firebaseCollectionName)
          .doc(integrationId)
          .delete();
    } catch (e) {
      throw Exception(
        "Error deleting telegram bot integration: ${e.toString()}",
      );
    }
  }

  Future<void> update({
    required TelegramBotIntegration integration,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to update this data.");
    }
    try {
      await firestore
          .collection(firebaseCollectionName)
          .doc(integration.id)
          .update(
            TelegramBotIntegrationFirebaseTransformer.toJson(integration),
          );
    } catch (e) {
      throw Exception(
        "Error updating telegram bot integration: ${e.toString()}",
      );
    }
  }

  Future<void> updateCustomFields({
    required String integrationId,
    required Map<String, dynamic> updatedData,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to update this data.");
    }
    try {
      await firestore
          .collection(firebaseCollectionName)
          .doc(integrationId)
          .update(updatedData);
    } catch (e) {
      throw Exception(
        "Error updating telegram bot integration: ${e.toString()}",
      );
    }
  }

  Future<bool> exists({
    required String integrationId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }
    try {
      DocumentSnapshot<Map<String, dynamic>> snapshot =
          await firestore
              .collection(firebaseCollectionName)
              .doc(integrationId)
              .get();

      return snapshot.exists;
    } catch (e) {
      throw Exception(
        "Error checking telegram bot integration existence: ${e.toString()}",
      );
    }
  }
}
